File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    dxdy = ((0,1),(1,0),(0,-1),(-1,0))
    N, M = len(maps), len(maps[0])
    visited = [[False] * M for _ in range(N)]
    q = deque([[0,0,1]])
    
    while q:
        x,y,cnt = q.popleft()
        if [x,y] == [N-1,M-1]: return cnt
        for dx,dy in dxdy:
            _x,_y = x+dx,y+dy
            if not (0 <= _x < N and 0 <= _y < M) or maps[_x][_y] == 0 or visited[_x][_y]: continue
            q.append([_x,_y,cnt+1])
            visited[_x][_y] = True

    return -1

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[][] dxdy = {{0,1}, {1,0}, {0,-1}, {-1,0}};
    public int solution(int[][] maps) {
        int N = maps.length, M = maps[0].length;
        Node node;
        Deque<Node> q = new ArrayDeque<>();
        q.addLast(new Node(0,0));
        
        while (!q.isEmpty()) {
            node = q.removeFirst();
            if (node.x == N-1 && node.y == M-1) return maps[node.x][node.y];
            for (int[] d: dxdy) {
                int x = node.x + d[0], y = node.y + d[1];
                if (!(0 <= x && x < N && 0 <= y && y < M) || maps[x][y] == 0) continue;
                if (maps[x][y] > 1) continue;
                maps[x][y] = maps[node.x][node.y]+1;
                q.addLast(new Node(x,y));
            }
        }
        return -1;
    }
    class Node {
        int x,y;
        Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}